### PR TITLE
#20504 move contentlet

### LINKED ISSUE #20504: Create a "Bulk Move" Action for Content
We need to create a move bulk action on the content search screen.  This action should be available like the other bulk actions are and should allow a user to select host or folder to which they would like to move the content.

Screen Shot 2021-06-08 at 1 43 33 PM

Once they have selected the host/folder, they can hit the button to "Apply" the move and we would send the selected list of contents \(as a lucene query\) to an endpoint that would then perform the "move". Just like the bulk workflow actions, the endpoint will produce a report of the contents that were successfully moved.  We should be able to use the new BulkContentUpdate endpoint to effect the move.

Once moved, the important thing is that the moved contents will now inherit the permissions from where they have been moved to - if they were moved to the system\_host, they will inherit permissions from the content type.

Adding this this should take care of issues #20196 and #14736 and give customers a way update the permissions on any given content type.

### LINKED ISSUE #20549: Create an endpoint to search for folders
The endpoint will return all the possible folders path that the are like the passed value.
We need to respect user permissions, so do not should folders that the user does not have View permissions.

E.g On How the endpoint should behave
```
SiteBrowser Tree:
default
	folder1
		subfolder1
		subfolder2
                 testsubfolder3
	folder2
		subfolder1
	testfolder3

default_copy
	folder1_copy
		subfolder1_copy
		subfolder2_copy
                 testsubfolder3_copy
	folder2_copy
		subfolder1_copy
	testfolder3_copy
```

Value Sent | Expected Result
------------ | -------------
//default/ | folder1, folder2, testfolder3
//default/fol | folder1, folder2
//default/fol/ | Nothing
//default/bla | Nothing
//default/folder1/ | subfolder1, subfolder2, testsubfolder3
//default/folder1/s | subfolder1, subfolder2
//default/folder1/b | Nothing
/ | folder1,folder2,testfolder3,folder1_copy,folder2_copy,testfolder3_copy
/f | folder1,folder2,folder1_copy,folder2_copy
/folder1/ | subfolder1, subfolder2, testsubfolder3,folder1_copy,folder2_copy,testfolder3_copy
/folder1/s | subfolder1, subfolder2,subfolder1_copy, subfolder2_copy
f | folder1,folder2,folder1_copy,folder2_copy
folder1/ | subfolder1, subfolder2, testsubfolder3,folder1_copy,folder2_copy,testfolder3_copy
folder1/s | subfolder1, subfolder2,subfolder1_copy, subfolder2_copy

Endpoint:
**POST {{serverURL}}/api/v1/folder/byPath**
 Body: 
 ```
 {
    "path": "//default/folder1/s"
}
```
Return:
List of FolderSearchResultView (path and hostname).

### PR DESCRIPTION
**Pull request by jdotcms**
This PR, adds the ability to:
1) Search folders (on Rest and Angular)
2) Move contentlet from some place to another:
   - 2.1 introduce a move method on the ContentletAPI
   - 2.2 introduce an actionlet for be used on workflow
   - 2.3 introduce changes to be runnable on wf bulk actions (REST AND UI)

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 1]
```diff
@@ -347,6 +350,81 @@ public Contentlet find(String inode, User user, boolean respectFrontendRoles) th
...
+
+            throw new DoesNotExistException("The identifier does not exists: " + contentlet.getIdentifier());
+        }
+
+        identifier.setHostId(host.getIdentifier());
+        identifier.setParentPath(folder.getPath());
+
+        Try.of(()->APILocator.getIdentifierAPI().save(identifier))
+                .getOrElseThrow(e -> new DotRuntimeException(e));
+
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:411]: Should we also update the host/folder values in the passed in `contentlet` as they are now stale?  I'm not sure.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 2]
```diff
@@ -347,6 +350,81 @@ public Contentlet find(String inode, User user, boolean respectFrontendRoles) th
...
+        HibernateUtil.addCommitListener(identifier.getId(), new FlushCacheRunnable() {
+            @Override
+            public void run() {
+                CacheLocator.getContentletCache().remove(contentlet.getInode());
+            }
+        });
+
+        if (ThreadContextUtil.isReindex()) {
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:419]: is the `if (ThreadContextUtil.isReindex()) {` needed?  It should just call `indexAPI.addContentToIndex(contentlet, false));`, no?

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/actionlet/MoveContentActionlet.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @Override
+    public String getName() {
+        return "Move";
+    }
+
+    @Override
+    public String getHowTo() {
+        return "This actionlet will move the content to a new path.  If the path is not set on the actionlet, will pop up a message to select the new destiny";
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/actionlet/MoveContentActionlet.java:40]: "If the path is not set on this actionlet, dotCMS will allow a user to select a destination".

**Review by wezell — CHANGES_REQUESTED** [review]
See some comments.  Also, needs tests :)

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java (hunk 1 above)
Comment (wezell) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:411]: Also, should we update the `version_ts` on the `contentlet_version_info` table?  If we don't do that, then the content won't be repushed after its been moved.

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java [hunk 4]
```diff
@@ -152,4 +153,93 @@ private final FolderView getFolders(final Folder folder, final User user){
...
+            throws DotSecurityException, DotDataException {
+        final List<FolderSearchResultView> subFolders = new ArrayList<>();
+
+        if(pathToSearch.lastIndexOf(StringPool.FORWARD_SLASH) == 0){ //If there is only one / we need to search the subfolders under the host(s)
+            if(UtilMethods.isSet(siteId)) {
+                subFolders.addAll(findSubfoldersUnderHost(siteId,pathToSearch,user));
+            } else{
+                final List<Host> siteList = APILocator.getHostAPI().findAll(user,false);
+                for(final Host site : siteList){
+                    if(subFolders.size() < 20) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:175]: this should be a config property

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java [hunk 5]
```diff
@@ -152,4 +153,93 @@ private final FolderView getFolders(final Folder folder, final User user){
...
+                        subFolders.addAll(findSubfoldersUnderHost(site.getIdentifier(), pathToSearch, user));
+                    }
+                    continue;
+                }
+            }
+
+        } else {
+            if(UtilMethods.isSet(siteId)) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:183]: I do not see the diff between the if or the else, take a look

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java [hunk 6]
```diff
@@ -152,4 +153,93 @@ private final FolderView getFolders(final Folder folder, final User user){
...
+    private List<FolderSearchResultView> findSubfoldersUnderHost(final String siteId,
+            final String pathToSearch, final User user)
+            throws DotSecurityException, DotDataException {
+        final List<FolderSearchResultView> subFolders = new ArrayList<>();
+
+        final Host host = APILocator.getHostAPI().find(siteId, user, false);
+        final List<Folder> subFoldersOfRootPath = APILocator.getFolderAPI()
+                .findSubFolders(host, user, false);
+        subFoldersOfRootPath.stream().filter(f -> f.getPath().startsWith(pathToSearch))
+                .limit(20).forEach(
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:212]: this should be config value

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:212]: rename f to folder

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java [hunk 7]
```diff
@@ -152,4 +153,93 @@ private final FolderView getFolders(final Folder folder, final User user){
...
+     * Will find the subfolders living directly under the host passed and the last valid folder (spliting the pathToSearch by the last '/').
+     * And filter the results by what is left after the last '/'.
+     */
+    private List<FolderSearchResultView> findSubfoldersUnderFolder(final String siteId,
+            final String pathToSearch, final User user)
+            throws DotSecurityException, DotDataException {
+        final List<FolderSearchResultView> subFolders = new ArrayList<>();
+        final Host host = APILocator.getHostAPI().find(siteId, user, false);
+
+        final int last_slash_pos = pathToSearch.lastIndexOf("/");
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:228]: wrong name convention

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:228]: replace by StringPool.FORWARD_SLASH

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderResource.java [hunk 8]
```diff
@@ -139,6 +145,97 @@ public final Response loadFolderAndSubFoldersByPath(@Context final HttpServletRe
...
+
+        if(path.startsWith(StringPool.DOUBLE_SLASH)){
+            //Removes // to search the host
+            path = path.startsWith(StringPool.DOUBLE_SLASH) ? path.substring(2) : path;
+            final String sitePath = path.split(StringPool.FORWARD_SLASH,2)[0];
+            final Host site = APILocator.getHostAPI().findByName(sitePath, user,false);
+            if(null == site){
+                throw new DoesNotExistException(String.format(" Couldn't find any host with name `%s` ",sitePath));
+            }else{
+                siteId = APILocator.getHostAPI().findByName(sitePath, user,false).getIdentifier();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderResource.java:229]: why ask again for the site, would be not just site.getIdentifier

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java (hunk 6 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:212]: Talked with Will about this, no right now.

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java (hunk 4 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:175]: Talked with Will about this, no right now.

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java (hunk 7 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:228]: done

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java (hunk 5 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderHelper.java:183]: one is to search folders under the host (Root level), the other is to search folders under folders.

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderResource.java (hunk 8 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderResource.java:229]: done

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 9]
```diff
@@ -703,6 +704,89 @@ public void testCheckInArchivedPageShouldThrowDotDataException() {
 
     }
 
+    /**
+     * Method to test: {@link ESContentletAPIImpl#move(Contentlet, User, Host, Folder, boolean)}
+     * Given Scenario: sends a null host and folder path
+     * ExpectedResult: The method should not throw a {@link IllegalArgumentException}
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:710]: Please update the doc of all tests included here, as it seems not to be totally related to what tests do or expect as result

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/MoveContentActionletTest.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        }
+
+        if (null != schemeStepActionResult) {
+
+            cleanScheme(schemeStepActionResult.getScheme());
+        }
+        cleanupDebug(MoveContentActionletTest.class);
+    } // cleanup
+
+    public Contentlet saveCustomTestContentType() throws DotDataException, DotSecurityException {
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/MoveContentActionletTest.java:129]: why is this method public?

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderResource.java [hunk 11]
```diff
@@ -139,6 +144,134 @@ public final Response loadFolderAndSubFoldersByPath(@Context final HttpServletRe
         return Response.ok(new ResponseEntityView(folderHelper.loadFolderAndSubFoldersByPath(siteId,path, user))).build(); // 200
     }
 
+    /**
+     * This endpoint is to retrieve subfolders of a given path,
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/folder/FolderResource.java:148]: great doc!

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 9 above)
Comment (jdotcms) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:710]: actually the description is accurate which the test

**Review by nollymar — APPROVED** [review]

**Pull request merged**